### PR TITLE
[PD] fix several hole bugs

### PR DESCRIPTION
**Pull request by donovaly**
- fix bug B reported here: https://forum.freecadweb.org/viewtopic.php?f=3&t=54408&sid=91194a40335be7962781e0aa7958a760
The problem was that we have normed screw head cuts. These values can be overridden but we must store the info about the overriding. Why? - because when you have e.g. made a custom change to a normed countersink and then change to another countersink norm, you would either not get the values defined in the norm or you get these values but loose e.g. your depth settings

- fix bug D reported here:
https://forum.freecadweb.org/viewtopic.php?f=3&t=54408&sid=91194a40335be7962781e0aa7958a760
The ReadOnly state of HoleCutCountersinkAngle was not correctly set for all cases

- fix bug C reported here: https://forum.freecadweb.org/viewtopic.php?f=3&t=54408&sid=91194a40335be7962781e0aa7958a760
There were 2 issues: the diameter widget had a fixed maximum and there was a superfluous spacer leading to unnecessary whitespace

- fix wrong clearance hole diameters when we have a UTS hole
We only had metric clearance holes defined and we calculated the UTS ones. However, there is a norm defining UTS clearances and since these do not follow a formula all UTS clearance holes are not correct.
This PR implements the norm ASME B18.2.8 to get rid of the calculation, this will only be used as fallback.

- fix tab order of the hole dialog

- fix this compiler warning: "C26451 Arithmetic overflow"
  (the code for the clearance hole diameters was anyway horrible and unnecessary)

- just a trifle: change a function name to fit into the naming scheme

Note for the merger: if possible, either please first only merge PR #4134 or this one, but not both the same day. They depend on each other, and I will have to adapt the second PR once the first is in.

### CODE REVIEW

**Comment by davidosterberg** [comment]
**B) (Custom contersink parameters are ignored)**

Fixed!

**C) (whole text is not visible in the hole diameter spinbox)**
All the text is now visible. Main annoyance is fixed :+1: . 
Do note that there is still a lot of whitespace that seems to be related to the choice of column and colspan of taper angle spinbox at the bottom of the dialog (see image) 

image

**D) (Angle spinbox is somethimes disabled when it should not be)**
This appears to be fixed.

File: src/Mod/PartDesign/Gui/TaskHoleParameters.ui
Comment (davidosterberg) [inline src/Mod/PartDesign/Gui/TaskHoleParameters.ui:189]: This spinbox causes a column of whitespace to appear for me (also in Master, not caused by this PR).
Simplest would be to constrain the spinbox max width to the same value as the ThreadClass combobox.
image

**Review by davidosterberg — CHANGES_REQUESTED** [review]
Nice! It is looking good. Only a few discussion points to resolve.

File: src/Mod/PartDesign/App/FeatureHole.cpp
Comment (donovaly) [reply src/Mod/PartDesign/App/FeatureHole.cpp:819]: > The new formula will reproduce the old values for whole numbers which is good. However, the formula will not give the same results for fractional values.

I know. But this formula is arbitrary and we only use it when we don't have a normed value, so only for the rather exotic M2.2, M9, M11 and M17. And the rounding error is negligible.

Concerning the clearance holes, the ones for the UTS profiles are not yet implemented.
But this is another issue. I cannot change too many things at once. Meanwhile I got the ASME 18.2.8 where the clearances are defined, I can do this as soon as this PR is in.

File: src/Mod/PartDesign/Gui/TaskHoleParameters.ui
Comment (donovaly) [reply src/Mod/PartDesign/Gui/TaskHoleParameters.ui:189]: > This spinbox causes a column of whitespace to appear for me

I have no clues why you get a different output than I do. Here is what I get with this PR:

FreeCAD_LXXqyVid2G

I checked the tapered angle field and it has the same settings than the dimension combobox and thus bot appear for me perfectly aligned, also when I change the width of the dialog in FC.
However, since you also see this in master this PR does not make it worse, in contrary, it fixes other UI issues for you.

File: src/Mod/PartDesign/Gui/TaskHoleParameters.ui
Comment (donovaly) [reply src/Mod/PartDesign/Gui/TaskHoleParameters.ui:23]: > Tab order is all screwy due to the random order of the rows and columns.

As I rote you can ignore the order of rows and columns, Qt designer does this automatically it has no effect. And I don't see why we must have them sorted, when Qt's own designer disorders them on every change.

The tab order however, is also incorrect in master. I fixed this now.

File: src/Mod/PartDesign/App/FeatureHole.cpp
Comment (davidosterberg) [reply src/Mod/PartDesign/App/FeatureHole.cpp:819]: Ok, old formula gives nicer values. The fractional ones will look strange on a drawing. I guess that was the original idea. And this will break old models. But it is up to you. Surely there is a way to get rid of the compiler warning by explicitly casting back to double. 
Your call! As long as the consequences are understood.

File: src/Mod/PartDesign/Gui/TaskHoleParameters.ui
Comment (davidosterberg) [reply src/Mod/PartDesign/Gui/TaskHoleParameters.ui:23]: I prefer if I can see what has changed at a glance by looking at the diff. And the tab order will be decent automatically. Running the script takes a minute, why not do it.

File: src/Mod/PartDesign/Gui/TaskHoleParameters.ui
Comment (davidosterberg) [reply src/Mod/PartDesign/Gui/TaskHoleParameters.ui:189]: Of course this is not blocking this PR. I was only hoping it could be fixed too.

File: src/Mod/PartDesign/App/FeatureHole.cpp
Comment (donovaly) [reply src/Mod/PartDesign/App/FeatureHole.cpp]: > And this will break old models.

There is no breakage, only for one special case the diameter will be slightly different. The only affected size is M2.2 and the change is < 0.6 %. For all other sizes we use the norm or the error is zero (e.g. M17).
However, the rounding was as arbitrary as the formula itself - it lead to wrong values for all UTS holes.

I fixed the missing UTS clearance hole definitions not too. Now we take the values given in the norm ASME B18.2.8. This changes the the dimameters of existing holes of course, but now we do it right.

File: src/Mod/PartDesign/Gui/TaskHoleParameters.ui
Comment (donovaly) [reply src/Mod/PartDesign/Gui/TaskHoleParameters.ui:23]: > I prefer if I can see what has changed at a glance by looking at the diff.

This is XML code, not designed to be human-readable. So also with sorted column names it won't become readable in mist cases.

> Running the script takes a minute, why not do it.

Because this costs extra time and I would have to do this every time I make a small change in Qt's designer. And it does not lead to the result you like to get. Change for example a grid layout to another layout. The resulting dialog can look almost the same but the diff is horrible. So please ignore the XML diffs. When they would be important, Qt would have designed the designer to maintain human-readable diffs.

File: src/Mod/PartDesign/App/FeatureHole.cpp
Comment (davidosterberg) [reply src/Mod/PartDesign/App/FeatureHole.cpp]: Sounds good. Any dimensional change should of course be done carefully, as we dont know if/how somebody is dependent on this diameter. That being said, if it is wrong we should fix it. Changing to better standard compliance sounds justified in this case.

File: src/Mod/PartDesign/Gui/TaskHoleParameters.ui
Comment (davidosterberg) [reply src/Mod/PartDesign/Gui/TaskHoleParameters.ui:23]: Terrible argument. The xml is perfectly human readable, by design. The dialog was sorted before so why should it be unsorted now. With the randomized order way you have no control over what has been unintentionally changed in the XML. 
Also, when coding, it is much faster to look in the XML to find out what a widgets name is. It is easy if the items are sorted. The tab order will be decent automatically without having to specify it manually. Lastly, when later you want to find what commit introduced a certain change (git blame). Changing the order every time the dialog is changed will make git blame not work.
These are the advantages of sorting the items. What is the advantages of not sorting?

Comment (donovaly) [reply src/Mod/PartDesign/Gui/TaskHoleParameters.ui:23]: > The dialog was sorted before so why should it be unsorted now.

Again, I am not to blame for design decisions made by Qt. I use their official tool to modify .ui files. Can you please report this to the Qt guys? It cannot be a solution that everyone has to code its own tool to keep the sorting.
And please discuss this in the forum because if we set up the rule to override QT Designer's sorting, every developer must be aware and know how to do this. Please CC there abdullah and wmayer who recently triggered the same discussion.

---------
(Personally, for me XML is in general not readable. To see the widget name, I can just click on the widget in designer and see its name. Opening the XML with a text editor and then searching costs just too much time.)

File: src/Mod/PartDesign/Gui/TaskHoleParameters.ui
Comment (davidosterberg) [reply src/Mod/PartDesign/Gui/TaskHoleParameters.ui:189]: The Size combobox has max width = 140
image

The TaperANgle combobox has max with of 1600000
image

This is the root cause for sure.

File: src/Mod/PartDesign/Gui/TaskHoleParameters.ui
Comment (davidosterberg) [reply src/Mod/PartDesign/Gui/TaskHoleParameters.ui:23]: This is not a general issue, it is primarily an issue for the QGridLayout. You are not responsible for Qt designer output of course. You are responsible for the Diff. I have given you the tool. 

I don't mind discussing it in the forum. Would be good to have general guidance. As an example of what I meant with "unintentional changes" to the ui that qt designer  may do. I opened the ui file and saved it. Qt designer 5.14.2 added native="true" and stdset="0" on many elements. Am I not responsible for whatever that change will do to the user experience?  A very strange position to take.

```
-    <widget class="Gui::PrefQuantitySpinBox" name="TaperedAngle">
+    <widget class="Gui::PrefQuantitySpinBox" name="TaperedAngle" native="true">
 ...
-     <property name="minimum">
+     <property name="minimum" stdset="0">
       <double>0.000000000000000</double>
```

We can discuss it further. And perhaps you will have your way. Or you could just apply this change, have a nice diff and clean git history. And we can move on.

File: src/Mod/PartDesign/App/FeatureHole.cpp
Comment (davidosterberg) [inline src/Mod/PartDesign/App/FeatureHole.cpp:853]: Having a helper function for this operation would reduce a lot of code repetition.

I.e  you could replace this with:

diameter = tableLookup(metricHoleDiameters, MatrixRowSizeMetric, diameter * 1.1);

**Review by davidosterberg — COMMENTED** [review]
Good additions in latest commit no severe issues. I request the introduction of a helper fuction to reduce code repetition. And there Still some unresolved discussion from the initial commit.

File: src/Mod/PartDesign/Gui/TaskHoleParameters.ui
Comment (davidosterberg) [reply src/Mod/PartDesign/Gui/TaskHoleParameters.ui:23]: https://forum.freecadweb.org/viewtopic.php?f=10&t=54503

File: src/Mod/PartDesign/App/FeatureHole.cpp
Comment (donovaly) [reply src/Mod/PartDesign/App/FeatureHole.cpp:853]: > Having a helper function for this operation would reduce a lot of code repetition.

I tried this now but did not gain shorter code. I still have to check 3 different cases for metric holes and UTS and the lookup is different for metric and UTS.

So if you have a clever proposal, please post it.

File: src/Mod/PartDesign/Gui/TaskHoleParameters.ui
Comment (donovaly) [reply src/Mod/PartDesign/Gui/TaskHoleParameters.ui:189]: > This is the root cause for sure.

This is strange because but have the policy Minimum for the width. So despite the tapered angle can run wider it does not need to.
However, I removed now the 140 px limit for the comboboxes. Please test if this fixes the issue for you.

Comment (chennes) [reply src/Mod/PartDesign/Gui/TaskHoleParameters.ui:189]: I can't speak to whether this fixes it for @davidosterberg, but mine looks like yours now, @donovaly. Perhaps a later PR could clean up what still looks to me to be some extra space in that column. At least in English, "Countersink Angle" is the longest thing in it, and has quite a bit of extra space after it.

File: src/Mod/PartDesign/App/FeatureHole.cpp
Comment (chennes) [inline src/Mod/PartDesign/App/FeatureHole.cpp:453]: I suggest a clarifying note here: I believe that standard actually lists a min and max hole size for each case, in inches. So if you used something like the Amesweb calculator to get the single metric values you might state that here, since it's not _quite_ exactly per the standard itself. And for the record I did double-check them, and they do match the Amesweb results.

File: src/Mod/PartDesign/App/FeatureHole.cpp
Comment (chennes) [inline src/Mod/PartDesign/App/FeatureHole.cpp:885]: In keeping with the error handling in the rest of this file, maybe consider:
```suggestion
                throw Base::IndexError("Thread fit out of range");
```

File: src/Mod/PartDesign/App/FeatureHole.cpp
Comment (chennes) [inline src/Mod/PartDesign/App/FeatureHole.cpp:935]:
```suggestion
                throw Base::IndexError("Thread fit out of range");
```

File: src/Mod/PartDesign/App/FeatureHole.cpp
Comment (chennes) [inline src/Mod/PartDesign/App/FeatureHole.cpp:957]:
```suggestion
                throw Base::IndexError("Thread fit out of range");
```

**Review by chennes — APPROVED** [review]
This compiles, runs, and appears to fix the issues stated in the description. Those I tested are:

- fix bug B reported here: https://forum.freecadweb.org/viewtopic.php?f=3&t=54408&sid=91194a40335be7962781e0aa7958a760

Tested, verified fixed.

- fix bug D reported here:
    https://forum.freecadweb.org/viewtopic.php?f=3&t=54408&sid=91194a40335be7962781e0aa7958a760

Tested, verified fixed.

- fix bug C reported here: https://forum.freecadweb.org/viewtopic.php?f=3&t=54408&sid=91194a40335be7962781e0aa7958a760

Tested: at least partially fixed, though there may still be room to remove additional whitespace from the dialog.

- fix wrong clearance hole diameters when we have a UTS hole

I double-checked the values used in this PR against the calculator at https://amesweb.info/Screws/clearance-hole-calculator.aspx and they match.

- fix tab order of the hole dialog

I don't know what the old tab order was, but the current tab order seems correct to me.

Overall this PR looks good to me. I've suggested unifying the error handling with the rest of the file to use `throw()` instead of `assert()` but it's a very minor issue considering that code should never be hit!

File: src/Mod/PartDesign/App/FeatureHole.cpp
Comment (davidosterberg) [inline src/Mod/PartDesign/App/FeatureHole.cpp:895]: Maybe I am showing my beginner status here. But are you actually comparing strings here? It looks like you are just comparing pointers. This means that the fallback will always be used. I.e the 1.1*diameter.

Comment (davidosterberg) [reply src/Mod/PartDesign/App/FeatureHole.cpp:895]: Well, never mind. It seems to be working.

File: src/Mod/PartDesign/App/FeatureHole.cpp
Comment (davidosterberg) [reply src/Mod/PartDesign/App/FeatureHole.cpp:853]: At least this saves some repetion. And it will be useful in the future. 
https://github.com/donovaly/FreeCAD/pull/2

It it a pity it does not work for the UTS tables, as they use a struct instead of a pure double array.

File: src/Mod/PartDesign/App/FeatureHole.cpp
Comment (donovaly) [reply src/Mod/PartDesign/App/FeatureHole.cpp:453]: Thanks for the comment. Yes, the standards defines actually a drill bit number (that is in turn standardized in another ASME norm). So the minimal clearance is the diameter of that drill bit. So we use this minimum as the theoretical exact diameter. This is also done in the ISO norm. That in reality, the resulting diameter of a drilling is wider, is clear and thus the ASME norm also defines the maximum allowed diameter, de-facto as tolerance.

I clarified this now in the comment.

File: src/Mod/PartDesign/App/FeatureHole.cpp
Comment (donovaly) [reply src/Mod/PartDesign/App/FeatureHole.cpp:885]: Thanks, implemented.

**Comment by donovaly** [comment]
> This compiles, runs, and appears to fix the issues stated in the description.

Thanks for testing. I made the minor changes you proposed.

@davidosterberg, let's please sort out further possible code optimizations and UI improvements to another PR. This one fixes several bugs and should go in first.

File: src/Mod/PartDesign/Gui/TaskHoleParameters.ui
Comment (davidosterberg) [reply src/Mod/PartDesign/Gui/TaskHoleParameters.ui:189]: Fine by me.

**Review by davidosterberg — CHANGES_REQUESTED** [review]
Functionally this PR compiles, runs, and fixes the mentioned bugs. The one change I must insist on is that the UI file is sorted. As requested the issue has been brought to the forum. It is overwhelmingly clear that I am not in the minority when requesting that the UI file order is maintained. The opinion of @abdullahtahiriyo and @OpenBrain were clearly stated. And the opinion of @wwmayer can be inferred from previous discussions. Please make the change and I will gladly approve the PR.

File: src/Mod/PartDesign/Gui/TaskHoleParameters.ui
Comment (davidosterberg) [reply src/Mod/PartDesign/Gui/TaskHoleParameters.ui:189]: Yep. It is working for me now. 
image

**Comment by davidosterberg** [comment]
Here is the sorted file. Also all occurences of stdset="0" and native="true" have been removed. 
[TaskHoleParameters.zip](https://github.com/FreeCAD/FreeCAD/files/5845895/TaskHoleParameters.zip)

**Comment by donovaly** [comment]
> Functionally this PR compiles, runs, and fixes the mentioned bugs. The one change I must insist on is that the UI file is sorted.

It annoys me because it cost me now all in all 30 minutes just for this code cosmetics to sort it.
So now the PR should please you.

However, it is a no-go that I am punished just because I use the official Qt tool to modify .ui files and that I am on Windows and don't know Python.

**Comment by davidosterberg** [comment]
> It annoys me because it cost me now all in all 30 minutes just for this code cosmetics to sort it.
> So now the PR should please you.

I am sorry to be annoying. I am very thankful about the work you put in, and I am sorry that this took so much time. Hopefully it will be smoother in the future when the tools have improved. The result pleases me. But there is one mistake on my part.
The script forgot to put the top row back into the file (because it is in a different tree). Could you please add
`<?xml version="1.0" encoding="UTF-8"?>`
At the top of the UI file.

**Comment by donovaly** [comment]
> But there is one mistake on my part.
> The script forgot to put the top row back into the file (because it is in a different tree). Could you please add
> `<?xml version="1.0" encoding="UTF-8"?>`
> At the top of the UI file.

This was not the only issue, your script also converted all existing LF line endings to Windows CRLF line endings leading to the maximal possible diff ;-)
I fixed both issues now.

So the PR is now really ready to be merged.

**Review by davidosterberg — APPROVED** [review]

**Comment by davidosterberg** [comment]
Great #4134 has been merged. @donovaly  can you fix the merge conflicts.

**Comment by donovaly** [comment]
> Great #4134 has been merged. @donovaly can you fix the merge conflicts.

It merges now but i see a lot of bugs I cannot explain right now - I am working on it.

File: src/Mod/PartDesign/App/FeatureHole.cpp
Comment (donovaly) [reply src/Mod/PartDesign/App/FeatureHole.cpp:900]: You are right and this explains the issues I see now.

**Comment by donovaly** [comment]
Phew, I think I got the problems:

- In updateHoleCutParams() the diameter must not be set again since it is set correctly in updateDiameterParam() according to the clearance hole norms (if it is a clearance hole of course).
- we must assure that the counterbore/sink diameter is not <= the hole diameter
- we must handle the case that we simply cannot propose a handy counterbore/sink diameter when there is no known norm defining this

So I made one further commit to this PR.

But since I cannot merge the different commits and I see it is really hard fora merger to follow this PR, I will make a new one based on today's master.

**Comment by donovaly** [comment]
I am closing this in favor of the "clean" PR #4337.

**Pull request closed without merging**
